### ISSUE TITLE
Memory bloat with AR find_in_batches

### ISSUE THREAD

**Issue by inkstak** [mperham/sidekiq#3229]
Hello, I'm investigating on a memory leak in Sidekiq. You can found a rails app [here](https://github.com/inkstak/memory_leak_test) to reproduce.

The gems are up to date
* rails 5.0.0.1
* sidekiq 4.2.5
* pg 0.19.0

I'm running this [job](https://github.com/inkstak/memory_leak_test/blob/master/app/jobs/mem_test_job.rb) to browse a large table with the Rails method `find_each`.
In this example, I try to call `GC.start` every 10 batches to reduce memory impact and then inspect the GC stats and current memory usage.

When the job is running in console, the heap size and the memory usage stabilize after few rounds.
```
GC.start                        After         Before           Last
 :count                           122            121            118
 :heap_available_slots        593,058        593,058        593,058
 :heap_live_slots             305,914        454,558        305,914
 :heap_free_slots             287,144        138,500        287,144
RSS: 125 MB
[...]
GC.start                        After         Before           Last
 :count                           125            124            122
 :heap_available_slots        593,058        593,058        593,058
 :heap_live_slots             305,914        455,057        305,914
 :heap_free_slots             287,144        138,001        287,144
RSS: 126 MB
```

Run the job through Sidekiq in production or with `config.eager_load = true` and you'll see the heap size and the RSS increase each round until the final blow up.
```
GC.start                        After         Before           Last
 :count                            90             89             88
 :heap_available_slots     12,703,267     12,703,267     12,082,901
 :heap_live_slots          12,655,023     12,703,122     12,034,158
 :heap_free_slots              48,244            145         48,743
RSS: 1000 MB
[...]
GC.start                        After         Before           Last
 :count                            92             91             90
 :heap_available_slots     13,324,859     13,324,859     12,703,267
 :heap_live_slots          13,275,993     13,324,440     12,655,023
 :heap_free_slots              48,866            419         48,244
RSS: 1.02 GB
```

* Performing the job through the rails console with `config.eager_load = false` doesn't cause a 
memory leak.
* Performing the job through the rails console with `config.eager_load = true` doesn't cause a memory leak.
* Performing the job through Sidekiq with `config.eager_load = false` doesn't cause a memory leak.
* Performing the job through Sidekiq with `config.eager_load = true` causes the leak.

**Comment by mperham** [comment]
If you comment out the Item batch call, does the memory leak go away?  In this case, Sidekiq is really a thin wrapper around a lot of Rails/AR/pg code.  It's likely the memory leak is in one of those three pieces.

**Comment by mperham** [comment]
Are these two items at odds?  One says "don't cache classes, reload them on every request", the other says "load all my code at startup".  I'm not sure how Rails will respond.

```ruby
  config.cache_classes = false
  config.eager_load = true
```

**Comment by inkstak** [comment]
```
  def run_test
    gc_start
    1000.times do |i|
      Item.limit(1000).to_a
      gc_start if i % 10 == 0
    end
  end
```

The memory quickly stabilizes after the first round.
So the problem should be in AR batches which may not be multithread-prof ?

**Comment by inkstak** [comment]
I first meet the leak in production where both config items are true.
I tried to reproduce it in development:
* enable only `eager_load` causes the leak
* enable only `cache_classes ` doesn't cause it.
* switching both and the leak is present

**Comment by mperham** [comment]
I suspect you are hitting the current buggy Rails 5 integration. Try Sidekiq 4.1.x.

**Comment by inkstak** [comment]
Using Sidekiq 4.1.x doesn't change anything.

I tried enqueing a Sidekiq worker instead of an ActiveJob job, and the bloat doesn't appear.
It's still present on my real app but not in this sample app, so I'll continue to investigate to find the gem in cause.

By the way, If you want to complete the wiki, the bloat appears when using the `.where`method.
It's limited to few bytes per requests, but keep growing.
```
1000.times do |i|
  Item.limit(1000).where('id > ?', i).to_a
  i += 1
  gc_start if i % 10 == 0
end
```

**Comment by inkstak** [comment]
I finally managed to stem the bloat by removing the [callback](https://github.com/rails/rails/blob/master/activejob/lib/active_job/railtie.rb#L25) initiated by ActiveJob to reload the code.

I guess AJ should not reload the app when `eager_load = true`.
I'll report the problem to the rails team and submit a PR soon

**Comment by shabbirsaifee92** [comment]
@inkstak Can you please explain what "removing the callback" exactly achieved?  We are facing a memory bloat as well with Rails 5.2.2

**Comment by inkstak** [comment]
Hi @shabbirsaifee92 

This is an old problem linked to a previous version of Rails. The link on the "callback" should point to  https://github.com/rails/rails/blob/v5.0.0.1/activejob/lib/active_job/railtie.rb#L25

But, as far as I remember, I was in the bad direction when I tried to solved this problem.
The problem was caused by find_in_batches and ActiveRecord caching and is fixed in Rails 5.2

**Comment by shabbirsaifee92** [comment]
@inkstak Thank you for the response :)
